[IV-21-22] Objetivo 4

arguellesm opened the pull request:
### Description

This PR:
- Starts implementing business logic via **blacklists** and a **Naive Bayes classifier algorithm**.
- Adds **tests** managed with Pytest for the class `Article`.


### Checklist

* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir una parte de la lógica de negocio que se
   esté desarrollando?
   _Yes, articles have now a `predict()` method that returns the probability
of that article of being misleading or fake._
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
   _[Milestone 0.2.0](https://github.com/arguellesm/faux/milestone/3)._
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x]  ¿Se han documentado las elecciones de biblioteca de aserciones y de
   *framework* de test?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?

arguellesm committed:
add data/*.csv to .gitignore

arguellesm committed:
first model: mnb

arguellesm committed:
create data directory

arguellesm committed:
add blacklist

arguellesm committed:
add model for predictions closes #12

arguellesm committed:
add tests for article closes #13

arguellesm committed:
add dependencies

arguellesm committed:
minor change

arguellesm committed:
add prediction capabilities improves #12

arguellesm committed:
document test framework choice

arguellesm committed:
add script to generate dataframe

arguellesm committed:
improve documentation

arguellesm committed:
add test to

arguellesm committed:
add test task

arguellesm committed:
update

arguellesm commented:
@JJ ready! 👀

JJ commented on data/generate-df.py:
```diff
@@ -0,0 +1,95 @@
+	if 'authors' in file.keys():
+		author = file['authors']
+
+	return file['text'], author
```
You should really use this in an OO way. You're simply returning an array. Clean code calls for using data structures for its purpose, and arrays were not made for this.

arguellesm replied:
I've modified it to make it cleaner, although this script is not part of the application it self, it's just a tool in case someone wants to replicate how the dataframe was created, as the version from FakeHealth has a different format.

JJ replied:
Best practices should be applied all across the project, not only in modules.

JJ commented on docs/dev_tools.md:
```diff
@@ -22,9 +22,10 @@ The main alternatives to Poetry are Pip and Pipenv:
 ### [Invoke](https://www.pyinvoke.org/)
 
-Invoke is our task runner of choice. It inherits a minimal yet 
+Invoke is our **task runner** of choice. It inherits a minimal yet 
```
Again, don't tell me your choice. Show me your best effort to choose it, and the criteria you have followed.

arguellesm replied:
The reason for choosing Invoke was that its community seems quite bigger. I find this useful for understanding and troubleshooting issues I might have in the future.

JJ commented on docs/dev_tools.md:
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
   _doit_ has quite a similar approach to Invoke and uses Python as 
```
... which is why it's used in Python.

JJ commented on docs/dev_tools.md:
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
+
+ ### [Pytest](https://docs.pytest.org/en/latest/)
+ 
+ Pytest is both our **test framework** and **assertion library** of
```
No, it's not. It barely has any assertions.

JJ replied:
> Just a few comments. I believe I improved the biggest issues pointed out before:
> 
>     1. Criteria are better know, with objective and measurable attributes.
> 
>     2. I broaden my range of possible tools searching for different assertion libraries that I didn't consider before.
> 
>     3. Exceptions are being tested now. There is one exception (no pun intended) to this: `ArticleDependencyError` will raise when a dependency is missing. This can't happen with the current Poetry setup, which is also used to run tests.
> 
>     4. The `predict` function is now split up into three different ones, that are also tested on their own: `rate_author`, `rate_source` and `rate_content`.
> 
>     5. Dev dependencies now include Invoke, Pytest and Pytruth.

Not a single one of the new commits refer to any issue. If they are addressing the PR comments, they should refer to them.

JJ commented on docs/dev_tools.md:
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
+ - Simplicity.
+
+When searching for test frameworks and assertion libraries, we searched
+for `python + [term]` and `python + [term] + stackoverflow`. These were
```
??? stackoverflow why?

JJ commented on docs/dev_tools.md:
```diff
@@ -42,3 +43,39 @@ Some alternatives are:
+  - [Unittest](https://docs.python.org/3/library/unittest.html)
+- Assertion libraries
+  - [Assertpy](https://pypi.org/project/assertpy/)
+  - [Unittest](https://docs.python.org/3/library/unittest.html)
```
Where have I seen this before? Ah, above. Where it does not belong. You can call it a test framework, but it's not by any means in the same class as pytest.

arguellesm replied:
I've modified this to include Unittest and the `assert` statement as the assertion library/tools 👍🏼

JJ commented on test/article_test.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def test_can_create_object_only_headline():
+    assert Article(headline='A simple headline')
```
This assertion does not belong to pytest. That means that you're not using pytest as an assertion library as you claim.

JJ replied:
Still don't belong to that. if you use truth, you should also use it here.

JJ reviewed (changes requested):
Besides all that's been said, there's no business logic tested here.

arguellesm committed:
improve generate-df.py

arguellesm committed:
improve reasons and add assertion library

arguellesm commented:
> Besides all that's been said, there's no business logic tested here.

I don't understand why. Business logic here includes predicting whether an article might be fake or not and to what extent, and I understand that's tested with:

```python
test_predict_returns_three_values()
test_predict_probability_between_0_and_1()
test_author_score_is_0()
test_source_score_is_0()
```

I guess I'm missing something but I don't know what (?)

Other than that, ready for review @JJ

JJ commented on docs/dev_tools.md:
```diff
@@ -2,30 +2,31 @@
 
-Invoke is our task runner of choice. It inherits a minimal yet 
-powerful philosophy from GNU Make and is broadly used among the 
+Invoke is our **task runner** of choice. It inherits a minimal yet
```
You should state your criteria before the tool you've chosen. If you don't, it looks like you're simply justifying what you had decided in advance. Also, not this objective, but the previous one. I should have told you so there.

arguellesm replied:
I've changed the structure of `dev_tools.md` in https://github.com/arguellesm/faux/pull/14/commits/7fa7ae2b94212e4514d805a78c706c74b4463edd to always list criteria first, then the alternatives I found and finally the option I ended up choosing.

The ones from the last objective are obviously done retroactively, but I'll keep it in mind in the future.

JJ commented on docs/dev_tools.md:
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
+
+ ### [Pytest](https://docs.pytest.org/en/latest/)
+
+ Pytest is both our **test framework** of
```
both what?

arguellesm replied:
fixed 👍🏼

JJ commented on docs/dev_tools.md:
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
+ Pytest is both our **test framework** of
+ choice. These were the criteria used in the decision:
+
+ - Well documented, preferably with a wide community behind it.
```
Again, criteria should be announced in advance, and should not be subjective, like "well documented". Are the others badly documented?

arguellesm replied:
I removed this criterion from the list 👍🏼

JJ commented on docs/dev_tools.md:
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
+ choice. These were the criteria used in the decision:
+
+ - Well documented, preferably with a wide community behind it.
+ - Already included in Python.
```
It's *not* included in Python.

arguellesm replied:
When listing criteria I don't expect the final tool to meet all of them. They're meant as a guide to give some alternatives preference over others.

JJ replied:
There's no test runner included in Python. No use to put up criteria that's going to be met by no tool.

JJ commented on docs/dev_tools.md:
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
+
+- Test frameworks
+  - [Pytest](https://docs.pytest.org/en/latest/)
+  - [Unittest](https://docs.python.org/3/library/unittest.html)
```
This is not a test runner. Pytest includes a test runner. I explained in class what's the concept of test runner, including the fact that it's able to parse TAP output.

JJ commented on docs/dev_tools.md:
```diff
@@ -34,11 +35,57 @@ Some alternatives are:
+In any case, failed assertions from Unittest provided added information
+on why the condition failed, which will be useful in the future.
+
+To allow simplicity when needed, we'll be also using the `assert`
```
How do you test exceptions then? Also, why don't you include possible exceptions in your code?

JJ commented on pyproject.toml:
```diff
@@ -9,6 +9,8 @@ license = "GPL-3.0"
 python = "^3.9"
 invoke = "^1.6.0"
 pylint = "^2.11.1"
+scikit-learn = "^1.0.1"
```
Is this used?

JJ replied:
Not clear why you need to pickle the model.

JJ commented on pyproject.toml:
```diff
@@ -9,6 +9,8 @@ license = "GPL-3.0"
 invoke = "^1.6.0"
 pylint = "^2.11.1"
+scikit-learn = "^1.0.1"
+pytest = "^6.2.5"
```
Is this a production or a dev dependency?

arguellesm replied:
I believe they're regular dependencies as they are always needed for the product to function properly.

JJ replied:
No, they're not. Testing is not a production operation, it's a dev operation.

JJ commented on src/article.py:
```diff
@@ -89,4 +92,29 @@ def id(self):
+
+        # check content
+        file = open(os.path.dirname(__file__)+'/../model/mnb.pkl', 'rb')
+        model = pickle.load(file)
```
This is not your own logic, it's a model you've picked up from elsewhere. You're not really testing your logic, but some other logic. This objective is about whitelist testing, and about testing your own business logic.

arguellesm replied:
I did the model myself as described in [model.ipynb](https://github.com/arguellesm/faux/blob/7fa7ae2b94212e4514d805a78c706c74b4463edd/model/model.ipynb) from  https://github.com/arguellesm/faux/pull/14/commits/131f57aac56d2e1b1bfbdbc55d39f625370d9df6. At the end of that notebook, the model is pickled to use it in the project.

JJ replied:
You need to do it *as part of the project*, with business logic that's part of the current minimally viable product. That issue is related to #12, which is not related to any HU whatsoever. If it's not advancing an US, it's not adding value to the customer.
Remember, this objective is not about testing. It's about advancing your project to a minimally viable product that includes business logic and can be, from now on, tested with the also included infrastructure.

JJ commented on src/article.py:
```diff
@@ -89,4 +92,29 @@ def id(self):
+            source_score = 1
+
+        # check content
+        file = open(os.path.dirname(__file__)+'/../model/mnb.pkl', 'rb')
```
Not testing for errors, and also including a literal when it should be a constant.

arguellesm replied:
changed 👍🏼

JJ reviewed (changes requested):
Besides the issues about justification, you should test your own logic, not something contained in a pickled model. If you need a simpler model for a MVP, create that model. But it's always about testing *your* code for *your* business logic.

arguellesm committed:
typo

arguellesm committed:
make model path global

arguellesm committed:
add and test custom exceptions

arguellesm committed:
improve dev tools structure and reasons

arguellesm commented:
@JJ ready for review

JJ commented on docs/dev_tools.md:
```diff
@@ -1,44 +1,128 @@
 ## Development tools and why
 
-### [Poetry](https://python-poetry.org/)
+### Dependency management
```
No one is requesting you to select a tool for dependency management. Most languages have just one, so it's not really worth the while.

JJ replied:
Again, not addressed. You should have merged to main all this part and eliminated it from here.

JJ commented on docs/dev_tools.md:
```diff
@@ -1,44 +1,128 @@
+them comfortably.
+
+These were the considered criteria:
+- Common within the Python community
```
None of these criteria are really measurable or have any relation to your own project.

arguellesm replied:
I changed this and tried to add ways of measuring them. They still don't have any particular relation to my project, but that's because I don't think the project needs anything too specific from the test runner or assertion library.

JJ replied:
Projects *always* need something specific, if only because they have been created at a specific point in time and should use the best practices available at that specific time.

JJ commented on docs/dev_tools.md:
```diff
@@ -1,44 +1,128 @@
+  tests need to be inside a class that inherits from 
+  `unittest.TestCase`.
+
+Even though Unittest is already included in Python, we opted
```
Not comparable, not by a long shot. pytest is a test runner with a few assertions (mainly to complement assert), unittest is a testing framework without a test runner. Also, there are other tests runners and assertion libraries.

JJ replied:
Not addressed.

JJ commented on src/article.py:
```diff
@@ -1,5 +1,35 @@
+    """
+    pass
+
+class CantLoadContentModelError(FileNotFoundError):
```
Not developing or testing this. It's essential that you test exceptions.

arguellesm replied:
Done.

JJ reviewed (changes requested):
Still some problems.
1. Try to establish some real criteria for choosing libraries. We talked about that yesterday in class.
2. Try to open up a bit your horizon of possible tools.
3. If you don't really test your own business logic, but an external model, you should at least develop exceptions and test them.
4. The "predict" function does several different things. It should have separated functions (separation of concerns) that should be tested independently.
5. Please try and understand the difference between production and development dependencies and use them accordingly.

arguellesm committed:
restrict dependencies due to errors

arguellesm committed:
unfold predict method

arguellesm committed:
removes getters and leading underscore from class attributes

arguellesm committed:
new exceptions tests

arguellesm committed:
updates dev dependencies

arguellesm committed:
adds invoke to dev dependencies

arguellesm committed:
adds pytruth

arguellesm committed:
moves pytruth to dev dependencies

arguellesm committed:
uses pytruth assertions

arguellesm committed:
adds documentation regarding the assertion library

arguellesm commented:
Just a few comments. I believe I improved the biggest issues pointed out before:

1. Criteria are better know, with objective and measurable attributes.
2. I broaden my range of possible tools searching for different assertion libraries that I didn't consider before.
3. Exceptions are being tested now. There is one exception (no pun intended) to this: `ArticleDependencyError` will raise when a dependency is missing. This can't happen with the current Poetry setup, which is also used to run tests.
4. The `predict` function is now split up into three different ones, that are also tested on their own: `rate_author`, `rate_source` and `rate_content`.
5. Dev dependencies now include Invoke, Pytest and Pytruth.

JJ commented:
⛹ Revisores → @Mapachana @Paszser @Parka015

arguellesm committed:
base container discussion closes #15

arguellesm committed:
add Dockerfile closes #16

arguellesm committed:
automate container build and push to dockerhub #17

arguellesm committed:
add invoke task to run tests inside container

arguellesm committed:
removes blank lines

JJ commented on data/README.md:
```diff
@@ -0,0 +1,26 @@
+### Processing 
+
+The original dataset is not in CSV format and has many fields we are not 
+currently intereted in. We've extracted the following fields of interest and
```
typo

JJ commented on data/README.md:
```diff
@@ -0,0 +1,26 @@
+The original dataset is not in CSV format and has many fields we are not 
+currently intereted in. We've extracted the following fields of interest and
+turned it into a single `.csv` file (which is not included in this repository).
+
```
At this point, you should have probably just modeled what's in that data and/or faked it. If you decide to process it, you should have created an issue for it, included in a milestone, and associated it to an US

arguellesm replied:
Addressed below

JJ commented on data/generate-df.py:
```diff
@@ -0,0 +1,104 @@
+import pandas as pd
```
Not a single line of this file is associated with an issue. Working with issues was essential since objective 2. Again, you don't need real data to test, test data can simply be fake data. If it's the structure you're looking for, you should have created an issue to model that.

arguellesm replied:
Addressed below

JJ commented on data/generate-df.py:
```diff
@@ -0,0 +1,104 @@
+		author = file['author']
+
+	if 'authors' in file.keys():
+		author = file['authors']
```
This is the consequence of not creating issues for modeling the problem domain. You're putting author and authors in the same output variable, when quite obviously one of them is singular and the other plural. You're also extracting information without any reference of what or how it's going to be used in your business logic. At this point in the project, it's essential that you understand, model, solve what's going or not going to be included as classes in your project.

arguellesm replied:
Addressed below

JJ commented on data/generate-df.py:
```diff
@@ -0,0 +1,104 @@
+	"""
+
+	# json fields
+	headline 	= []
```
All these variables are in singular, while they are initialized with an array. Clean code calls for "semantic" identifiers, that really point to what they hold or contain.

arguellesm replied:
Addressed below

JJ commented on data/generate-df.py:
```diff
@@ -0,0 +1,104 @@
+			continue
+
+		text.append(get_text(path))
+		author.append(get_author(path))
```
[`append`](https://towardsdatascience.com/append-in-python-41c37453400) appends a single item; if it's the result of the original "authors" field you either are going to add an array as a single item, or are going to add a (possibly comma) separated list of authors as a single one. Not clear what's the case here, but again this is a reason why you need to *always* add to the repo according to HUs/issues.

arguellesm replied:
It adds an array of authors. I've changed the variable name to `authors` to make it more descriptive.

JJ commented on docs/dev_tools.md:
```diff
@@ -1,44 +1,190 @@
+one. 
+
+
+### Test framework
```
We ask for two things (please check out the revised objective description):
* a *test runner*
* an *assertion library*

Talking about testing frameworks is misleading, because it can include everything, from mocking  libraries to runner-assertion-coverage combos.

arguellesm replied:
Changed 👍🏼

JJ commented on docs/dev_tools.md:
```diff
@@ -1,44 +1,190 @@
+When searching for test frameworks, we searched for `python + 
+[term]` and `python + [term] + stackoverflow`. These were the 
+main results:
+
```
Again, you shouldn't. You should have tried and understand first, as we told in class, what's a test runner, and what are assertion libraries.

JJ commented on src/article.py:
```diff
@@ -1,5 +1,35 @@
-import globals
+from src import globals, blacklists
```
Best practices dictate that the module should be in a directory with the name of the project. It should never be "src"

arguellesm replied:
Changed name to `faux` 👍🏼

JJ commented on src/article.py:
```diff
@@ -32,61 +62,99 @@ def __init__(self, headline='', author='', content='', topic='', source=''):
-            raise ValueError('Headline cannot be larger than ' + str(globals.HEADLINE_MAX_SIZE))
+            raise ArticleValueError('Headline cannot be larger than {}'.format(globals.HEADLINE_MAX_SIZE))
         
         if(globals.AUTHOR_MAX_LENGHT > len(author)):
```
typo. Also, don't understand why tests are not all here.

arguellesm replied:
All tests are in the `/test` directory. I didn't fix the typo because I couldn't find it.

JJ commented on tasks.py:
```diff
@@ -16,3 +16,12 @@ def check(c, format='colorized'):
+    Run Pytest tests.
+    """
+
+    c.run('python -m pytest test')
```
Why not simply pytest?

arguellesm replied:
I left it like that because it added the current path to `sys.path`, but I don't really use that. Changed to plain `pytest` 👍🏼

JJ reviewed (approved):
I wouldn't probably pass this if you had the chance to send it again. But you don't, so it's either hit or miss. Main positive point is to have tried and be original with the assertion library, and try to test the business logic (although, again, you need to use only your own code). There are many negative points that have been pointed out, and that will probably jeopardize the fulfillment of the rest of the objectives if you don't address them.

arguellesm committed:
documents CI tools choice #19

arguellesm committed:
adds github action for language version testing #20

arguellesm committed:
adds gitlab CI pipeline for unit testing

arguellesm committed:
updates min python allowed version

arguellesm committed:
documents CI actions and pipelines

arguellesm committed:
fix branch name on update_docker github action

arguellesm committed:
adds CI to iv.yaml

arguellesm committed:
fixes typo

arguellesm committed:
fixes misleading term: test runner instead of test framework

arguellesm committed:
changes module name to faux

arguellesm committed:
changes module import to faux

arguellesm committed:
removes assert statement and uses AssertThat

arguellesm committed:
changes way of invoking pytest

arguellesm committed:
changes variables names into more descriptive ones

arguellesm committed:
update README.md

arguellesm commented:
Addressing the issues in `generate_df.py`:

- This file shouldn't have been included here in the first place. It's a script to (sort of) turn a JSON file into a CSV dataframe that 
it's not a requirement for the project. However, I opted for including it for documentation and replication purposes, so that anyone could see where the `.csv` file came from. I now see several problems with this:
  - There is no need to have a 'proper dataset' at this stage and   I should have used that data as dummy or fake data as you suggest.
  - By doing so, there will be no point in documenting that, not to mention providing the means to replicate it.
  - The script was thought of as a tool to generate something else and be discarded later, not as an integral part of the project, and for that reason it shouldn't have been added to the repository.
  
- On top of that, there are many issues with the `generate_df.py` script and the original JSON dataset. It was difficult to extract meaningful data because some categories would have multiple names (like 'author' and 'authors'), some keys wouldn't be present for all instances, and many values were blank. This and the fact that the script wasn't initially meant to be shared led to a messy and not very thoughtful script. It was also a mistake to rush and force having "real" data, as this could have been easily avoided with fake data, as commented before.

- I've modified a few things from `generate_df.py` to answer some of your comments, although I believe the best option would be to delete the file and consider the model was trained on fake data. Even further, there should be no data to train the model on as the model should be replaced by another implementation of the business logic that didn't require external libraries.

**tl;dr** I've made some changes to `generate_df.py` to address your comments. Regardless, I believe the best alternative would be to consider the model was based on fake/dummy data and delete `generate_df.py`.

arguellesm committed:
changes CI in iv.yaml

arguellesm committed:
adds both CIs to iv.yaml

arguellesm committed:
changed pull request trigger type on github action

arguellesm committed:
reconsiders size and expands documentation

arguellesm committed:
Merge pull request #23 from arguellesm/objetivo-6

[IV-21-22]  Objetivo 6

arguellesm committed:
Merge branch 'objetivo-4' into objetivo-5

arguellesm committed:
Merge pull request #18 from arguellesm/objetivo-5

[IV-21-22] Objetivo 5
